client/image_pull.go

```diff
@@ -2,19 +2,50 @@ package client
 
 import (
 	"context"
+	"encoding/json"
+	"errors"
 	"io"
+	"iter"
 	"net/url"
 	"strings"
 
 	cerrdefs "github.com/containerd/errdefs"
 	"github.com/distribution/reference"
+	"github.com/moby/moby/client/pkg/jsonmessage"
 )
 
+type ImagePullResponse struct {
+	io.ReadCloser
+}
+
+// Messages decodes the response stream as a sequence of JSONMessages
+func (r *ImagePullResponse) Messages() iter.Seq2[*jsonmessage.JSONMessage, error] {
```

Comment:
I think we should make it more explicit -

```suggestion
func (r *ImagePullResponse) JSONMessages() iter.Seq2[*jsonmessage.JSONMessage, error] {
```

In future, I think we should have a better public interface than `JSONMessage` and we won't need to change the signaure of `Messages`.

Reply:
Also, perhaps worth passing a `ctx` so there's a way to cancel this without consuming all messages?